src/Components/Components/test/Routing/RouterTest.cs

```diff
@@ -38,6 +38,26 @@ async Task OnNavigateAsync(NavigationContext args)
             Assert.True(called);
         }
 
+        [Fact]
+        public async Task CanHandleSingleFailedOnNavigateAsync()
```

Comment:
Sorry for the fussiness, but would it be possible to consider doing this test without the mocking and the use of an artificial testability seam? (i.e., the `internal virtual` method)

Whenever a test uses mocking over an internal nonpublic API, each such test duplicates assumptions about internal implementation details. Over time this makes the production code harder to alter internally because refactoring (that doesn't touch public APIs) knocks over lots of tests. And then, reverse-engineering what the developer meant 18 months ago when chaining a set of Moq API calls is the worst!

Additionally in this case we have missed out on coverage of what `ThrowExceptionInRenderer` really does.

In this case I suspect it will be very straightforward to eliminate the "mock router" concept and just directly use `new Router()`. Then something like:

```cs
var testRenderer = new TestRenderer();
testRenderer.AssignRootComponentId(router);
```

And then after you do the thing to make it report an exception, you can read it directly from `testRenderer.HandledExceptions` and assert you got the desired exception. And no more need for the `internal virtual` 😉

Reply:
Note, I'm not asking for the other unit tests here that mock `Refresh` to be changed. I recognize it would be hard in that case (maybe possible, but it doesn't seem worth it). I'm just raising it here because I think the alternative is quite simple, unless I'm missing a hidden issue.

Reply:
Cool idea! I was opting for the quickest route here and forgotten that we already had a "mock" for the renderer in the form of the `TestRenderer` implementation.